File: packages/vaultkeeper/src/keys/manager.ts
/**
 * Key management for vaultkeeper, including generation, rotation, revocation,
 * and grace-period tracking.
 */

import * as crypto from 'node:crypto'
import { RotationInProgressError, SetupError } from '../errors.js'
import type { KeyMaterial, KeyState, KeyStateSnapshot } from './types.js'

/**
 * Manage cryptographic keys with rotation and grace-period semantics.
 *
 * @remarks
 * Whether a rotation is "in progress" is derived entirely from
 * {@link KeyManager.isInGracePeriod} — a grace period is active exactly while a
 * previous key is still usable. This makes the state serializable via
 * {@link KeyManager.snapshot}/{@link KeyManager.hydrate}, so the guard against a
 * second rotation survives a process restart (see `keys/storage.ts`) rather than
 * relying on a live in-process timer.
 * @internal
 */
export class KeyManager {
  #state: KeyState | undefined = undefined
  #gracePeriodTimer: ReturnType<typeof setTimeout> | undefined = undefined
  #gracePeriodExpiresAt: number | undefined = undefined

  /** Generate a new 32-byte key with a timestamp-based id. */
  generateKey(): KeyMaterial {
    const randomSuffix = crypto.randomBytes(4).toString('hex')
    return {
      id: `k-${String(Date.now())}-${randomSuffix}`,
      key: new Uint8Array(crypto.randomBytes(32)),
      createdAt: new Date(),
    }
  }

  /**
   * Initialize the manager with a freshly generated key.
   * Safe to call multiple times; subsequent calls are no-ops.
   */
  init(): Promise<void> {
    if (this.#state === undefined) {
      this.#state = { current: this.generateKey() }
    }
    return Promise.resolve()
  }

  /**
   * Replace the in-memory state with a persisted snapshot.
   *
   * An expired grace period in the snapshot is dropped: the previous key is
   * discarded and no grace period is restored. When the grace period is still
   * active, a fresh timer is armed for the remaining duration so the previous
   * key is eventually freed in long-running processes; correctness never depends
   * on that timer (grace is always re-checked against the wall clock).
   * @internal
   */
  hydrate(snapshot: KeyStateSnapshot): void {
    this.#clearGracePeriodTimer()

    if (
      snapshot.previous !== undefined &&
      snapshot.gracePeriodExpiresAt !== undefined &&
      Date.now() < snapshot.gracePeriodExpiresAt
    ) {
      this.#state = { current: snapshot.current, previous: snapshot.previous }
      this.#gracePeriodExpiresAt = snapshot.gracePeriodExpiresAt
      this.#armGracePeriodTimer(snapshot.gracePeriodExpiresAt - Date.now())
    } else {
      this.#state = { current: snapshot.current }
      this.#gracePeriodExpiresAt = undefined
    }
  }

  /**
   * Capture the current state as a serializable snapshot for persistence.
   * The previous key and grace expiry are included only while the grace period
   * is still active.
   * @internal
   */
  snapshot(): KeyStateSnapshot {
    const state = this.#requireState()
    const result: KeyStateSnapshot = { current: state.current }
    const previous = this.getPreviousKey()
    if (previous !== undefined && this.#gracePeriodExpiresAt !== undefined) {
      result.previous = previous
      result.gracePeriodExpiresAt = this.#gracePeriodExpiresAt
    }
    return result
  }

  /** Return the current (encryption) key. Throws if not initialized. */
  getCurrentKey(): KeyMaterial {
    const state = this.#requireState()
    return state.current
  }

  /**
   * Return the previous key if we are still inside a grace period,
   * otherwise `undefined`.
   */
  getPreviousKey(): KeyMaterial | undefined {
    const state = this.#requireState()
    // Grace is authoritative: even if a timer has not yet fired to clear the
    // reference, an expired grace period means the previous key is unusable.
    if (!this.isInGracePeriod()) {
      return undefined
    }
    return state.previous
  }

  /**
   * Find a key by its id, searching current then previous.
   * Returns `undefined` if the key is not found (or the previous key's
   * grace period has expired).
   */
  findKeyById(kid: string): KeyMaterial | undefined {
    const state = this.#requireState()
    if (state.current.id === kid) {
      return state.current
    }
    const previous = this.getPreviousKey()
    if (previous?.id === kid) {
      return previous
    }
    return undefined
  }

  /**
   * Rotate the current key: the current key becomes previous, a new key
   * becomes current. A grace-period timer is started; when it fires the
   * previous key is cleared automatically.
   *
   * @throws {RotationInProgressError} if a rotation is already underway (i.e. a
   *   previous key is still within its grace period).
   */
  rotateKey(gracePeriodMs: number): void {
    const state = this.#requireState()

    if (this.isInGracePeriod()) {
      throw new RotationInProgressError('A key rotation is already in progress')
    }

    // Cancel any stale timer (e.g. from a grace period that has already
    // elapsed) before starting a new rotation.
    this.#clearGracePeriodTimer()

    const newKey = this.generateKey()
    this.#state = { current: newKey, previous: state.current }
    this.#gracePeriodExpiresAt = Date.now() + gracePeriodMs
    this.#armGracePeriodTimer(gracePeriodMs)
  }

  /**
   * Emergency revocation: immediately clear the previous key and generate
   * a brand-new current key. Any in-flight grace period is cancelled.
   */
  revokeKey(): void {
    this.#clearGracePeriodTimer()
    this.#gracePeriodExpiresAt = undefined

    const newKey = this.generateKey()
    this.#state = { current: newKey }
  }

  /**
   * Return `true` while a rotation grace period is active (i.e. the previous
   * key is still accessible).
   */
  isInGracePeriod(): boolean {
    if (this.#gracePeriodExpiresAt === undefined) {
      return false
    }
    return Date.now() < this.#gracePeriodExpiresAt
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  #requireState(): KeyState {
    if (this.#state === undefined) {
      throw new SetupError('KeyManager has not been initialized — call init() first', 'KeyManager')
    }
    return this.#state
  }

  #armGracePeriodTimer(delayMs: number): void {
    this.#gracePeriodTimer = setTimeout(() => {
      if (this.#state !== undefined) {
        // Remove the previous key once the grace period elapses.
        this.#state = { current: this.#state.current }
      }
      this.#gracePeriodExpiresAt = undefined
      this.#gracePeriodTimer = undefined
    }, delayMs)

    // Allow the timer to be GC'd without keeping the process alive.
    if (typeof this.#gracePeriodTimer.unref === 'function') {
      this.#gracePeriodTimer.unref()
    }
  }

  #clearGracePeriodTimer(): void {
    if (this.#gracePeriodTimer !== undefined) {
      clearTimeout(this.#gracePeriodTimer)
      this.#gracePeriodTimer = undefined
    }
  }
}

File: crates/vaultkeeper-core/src/keys/manager.rs
//! KeyManager — handles key generation, rotation, and lookup.

use super::types::{KeyMaterial, KeyState};
use crate::errors::VaultError;
use crate::util::time;
use std::sync::atomic::{AtomicU64, Ordering};

/// Monotonic counter to ensure unique key IDs even within the same millisecond.
static KEY_COUNTER: AtomicU64 = AtomicU64::new(0);

/// Manages encryption keys and rotation lifecycle.
pub struct KeyManager {
    state: Option<KeyState>,
}

impl KeyManager {
    /// Create a new uninitialized KeyManager.
    pub fn new() -> Self {
        Self { state: None }
    }

    /// Initialize the key manager, generating a fresh key.
    pub fn init(&mut self) -> Result<(), VaultError> {
        let key = Self::generate_key()?;
        self.state = Some(KeyState {
            current: key,
            previous: None,
            rotated_at_ms: None,
            grace_period_ms: None,
        });
        Ok(())
    }

    /// Get the current (active) key.
    pub fn get_current_key(&self) -> Result<&KeyMaterial, VaultError> {
        self.state
            .as_ref()
            .map(|s| &s.current)
            .ok_or(VaultError::Other("KeyManager not initialized".to_string()))
    }

    /// Get the previous key (if in grace period).
    pub fn get_previous_key(&mut self) -> Option<&KeyMaterial> {
        self.expire_previous_if_needed();
        self.state.as_ref().and_then(|s| s.previous.as_ref())
    }

    /// Find a key by its ID. Returns the key and whether it is the current key.
    pub fn find_key_by_id(&mut self, kid: &str) -> Option<(&KeyMaterial, bool)> {
        self.expire_previous_if_needed();
        let state = self.state.as_ref()?;
        if state.current.id == kid {
            return Some((&state.current, true));
        }
        if let Some(ref prev) = state.previous
            && prev.id == kid
        {
            return Some((prev, false));
        }
        None
    }

    /// Rotate the current key. The old key becomes `previous` for the grace period.
    pub fn rotate_key(&mut self, grace_period_ms: u64) -> Result<(), VaultError> {
        self.expire_previous_if_needed();

        let state = self
            .state
            .as_mut()
            .ok_or(VaultError::Other("KeyManager not initialized".to_string()))?;

        if state.previous.is_some() {
            return Err(VaultError::RotationInProgress {
                message: "A key rotation is already in progress".to_string(),
            });
        }

        let new_key = Self::generate_key()?;
        let old_current = std::mem::replace(&mut state.current, new_key);
        state.previous = Some(old_current);
        state.rotated_at_ms = Some(time::now_millis());
        state.grace_period_ms = Some(grace_period_ms);

        Ok(())
    }

    /// Lazily expire the previous key if the grace period has elapsed.
    fn expire_previous_if_needed(&mut self) {
        let Some(state) = self.state.as_mut() else {
            return;
        };
        if state.previous.is_none() {
            return;
        }
        if let (Some(rotated_at), Some(grace_ms)) = (state.rotated_at_ms, state.grace_period_ms) {
            let now = time::now_millis();
            if now >= rotated_at + u128::from(grace_ms) {
                state.previous = None;
                state.rotated_at_ms = None;
                state.grace_period_ms = None;
            }
        }
    }

    /// Emergency key revocation — removes the previous key immediately.
    pub fn revoke_key(&mut self) -> Result<(), VaultError> {
        let state = self
            .state
            .as_mut()
            .ok_or(VaultError::Other("KeyManager not initialized".to_string()))?;
        state.previous = None;

        // Generate a new current key
        let new_key = Self::generate_key()?;
        state.current = new_key;

        Ok(())
    }

    /// Generate a new 32-byte key with a unique ID.
    fn generate_key() -> Result<KeyMaterial, VaultError> {
        let mut key_bytes = vec![0u8; 32];
        getrandom::fill(&mut key_bytes).map_err(|e| VaultError::Other(e.to_string()))?;

        let ts = time::now_millis();
        let seq = KEY_COUNTER.fetch_add(1, Ordering::Relaxed);
        let id = format!("k-{ts}-{seq}");

        Ok(KeyMaterial {
            id,
            key: key_bytes,
            created_at: time::now_secs(),
        })
    }
}

impl Default for KeyManager {
    fn default() -> Self {
        Self::new()
    }
}
